crates/ty_python_semantic/resources/mdtest/type_properties/satisfied_by_all_typevars.md

```diff
@@ -0,0 +1,158 @@
+# Constraint set satisfaction
+
+```toml
+[environment]
+python-version = "3.12"
+```
+
+Constraint sets exist to help us check assignability and subtyping of types in the presence of
+typevars. We construct a constraint set describing the conditions under which assignability holds
+between the two types. Then we check whether that constraint set is satisfied for the valid
+specializations of the relevant typevars. This file tests that final step: ensuring that a
+constraint set is satisfied given a list of inferable and non-inferable typevars.
+
+(Note that in this file, for ease of reproducibility, we explicitly list the typevars that are
+inferable in each `satisfied_by_all_typevars` call; any typevar not listed is assumed to be
+non-inferable.)
+
+## Inferable typevars
+
+When a typevar is in an inferable position, the constraint set only needs to be satisfied for _some_
+valid specialization. The most common inferable position occurs when invoking a generic function:
+all of the function's typevars are inferable, because we want to use the argument types to infer
+which specialization is being invoked.
+
+```py
+from typing import final, Never
+from ty_extensions import ConstraintSet, static_assert
+
+class Super: ...
+class Base(Super): ...
+class Sub(Base): ...
+
+@final
+class Unrelated: ...
+```
+
+If a typevar has no bound or constraints, then it can specialize to any type. For an inferable
+typevar, that means we just need a single type (any type at all!) that satisfies the constraint set.
+
+```py
+def unbounded[T]():
+    static_assert(ConstraintSet.always().satisfied_by_all_typevars(T))
+    static_assert(not ConstraintSet.never().satisfied_by_all_typevars(T))
+    static_assert(ConstraintSet.range(Never, T, Unrelated).satisfied_by_all_typevars(T))
+    static_assert(ConstraintSet.range(Never, T, Super).satisfied_by_all_typevars(T))
+    static_assert(ConstraintSet.range(Never, T, Base).satisfied_by_all_typevars(T))
+    static_assert(ConstraintSet.range(Never, T, Sub).satisfied_by_all_typevars(T))
+```
+
+If a typevar has an upper bound, then it must specialize to a type that is a subtype of that bound.
+For an inferable typevar, that means we need a single type that satisfies both the constraint set
+and the upper bound.
+
+```py
+def bounded[T: Base]():
+    static_assert(ConstraintSet.always().satisfied_by_all_typevars(T))
+    static_assert(not ConstraintSet.never().satisfied_by_all_typevars(T))
+    static_assert(ConstraintSet.range(Never, T, Super).satisfied_by_all_typevars(T))
+    static_assert(ConstraintSet.range(Never, T, Base).satisfied_by_all_typevars(T))
+    static_assert(ConstraintSet.range(Never, T, Sub).satisfied_by_all_typevars(T))
+
+    # This succeeds because T can specialize to Never
```

Comment:
I don't think I understand the why but I trust you on this

Reply:
> This tells me that it might help to change the structure of this file?

I did this, and added more explanatory comments for each example. Hopefully that helps clarify the logic a bit.